# Task 2

# Credit Card Fraud Detection

### Import Important Libraries

In [1]:
import numpy as np 
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.metrics import accuracy_score, precision_score



# import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))


# Importing the Datasets

In [2]:
train_df = pd.read_csv('/Users/sadgijakhar/Documents/Machine_Learning/Encryptix/Task_2/archive/fraudTrain.csv')
test_df = pd.read_csv('/Users/sadgijakhar/Documents/Machine_Learning/Encryptix/Task_2/archive/fraudTest.csv')

In [3]:
train_df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2  Sanchez      M      594 White Dale Suite 530  ...  42.1808 -112.2620   
3    White      M   9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4   Garcia      M              408 Bradley Rest  ...  38.4207  -79.4629   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [4]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1296675 entries, 0 to 1296674
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Unnamed: 0             1296675 non-null  int64  
 1   trans_date_trans_time  1296675 non-null  object 
 2   cc_num                 1296675 non-null  int64  
 3   merchant               1296675 non-null  object 
 4   category               1296675 non-null  object 
 5   amt                    1296675 non-null  float64
 6   first                  1296675 non-null  object 
 7   last                   1296675 non-null  object 
 8   gender                 1296675 non-null  object 
 9   street                 1296675 non-null  object 
 10  city                   1296675 non-null  object 
 11  state                  1296675 non-null  object 
 12  zip                    1296675 non-null  int64  
 13  lat                    1296675 non-null  float64
 14  long              

# Data Preprocessing

In [5]:
print("Checking Duplicate values:",train_df.duplicated().sum(), test_df.duplicated().sum())
print("Checking Null values:", train_df.isna().sum().sum(), test_df.isna().sum().sum())

Checking Duplicate values: 0 0
Checking Null values: 0 0


### Feature Selection

In [6]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()


# data preprocessing for Training Data
train_data = train_df[['gender', "category", "amt", "city", "job", "is_fraud"]]

train_data.loc[:, 'category'] = le.fit_transform(train_data["category"])
train_data.loc[:, 'city'] = le.fit_transform(train_data["city"])
train_data.loc[:, 'job'] = le.fit_transform(train_data["job"])
train_data.loc[:, 'gender'] = train_data["gender"].map({'M': 0, 'F': 1})

# data preprocessing for Testing Data
test_data = test_df[['gender',"category","amt","city", "job","is_fraud"]]

test_data.loc[:, 'category'] = le.fit_transform(test_data["category"])
test_data.loc[:, 'city'] = le.fit_transform(test_data["city"])
test_data.loc[:, 'job'] = le.fit_transform(test_data["job"])
test_data.loc[:, 'gender'] = test_data["gender"].map({'M': 0, 'F': 1})

In [7]:
train_data.head(10), test_data.head(10)

(  gender category     amt city  job  is_fraud
 0      1        8    4.97  526  370         0
 1      1        4  107.23  612  428         0
 2      0        0  220.11  468  307         0
 3      0        2   45.00   84  328         0
 4      0        9   41.96  216  116         0
 5      1        2   94.63  223  479         0
 6      1        3   44.54  351   29         0
 7      0        2   71.65  236  127         0
 8      1        9    4.27  474  375         0
 9      1        4  198.39  149  329         0,
   gender category     amt city  job  is_fraud
 0      0       10    2.86  157  275         0
 1      1       10   29.84   16  392         0
 2      1        5   41.28   61  259         0
 3      0        9   60.05  764  407         0
 4      0       13    3.19  247  196         0
 5      1        7   19.55   90  361         0
 6      1        5  133.93  117  455         0
 7      1       10   10.37  725  124         0
 8      0       12    4.37  503   13         0
 9      1   

In [8]:
X_train = train_data[['gender', "category", "amt", "city", "job"]]
Y_train = train_data["is_fraud"]

X_test = test_data[['gender', "category", "amt", "city", "job"]]
Y_test = test_data["is_fraud"]

## Logistic Regression
Logistic regression (L1 Regularization) is a statistical method for predicting binary classes. The outcome or target variable is binary, meaning it can take on two possible outcomes (0 or 1, true or false, etc.). It models the probability of the default class (usually 1) and uses a logistic function to produce outputs between 0 and 1.

In [9]:
lr_model = LogisticRegression(penalty="l1", solver="liblinear")
lr_model.fit(X_train, Y_train)

LogisticRegression(penalty='l1', solver='liblinear')

In [10]:
Y_pred = lr_model.predict(X_test)
accuracy = accuracy_score(Y_test, Y_pred)
print("Testing Accuracy Score:", accuracy)

Testing Accuracy Score: 0.9955013235106231


# Gaussian Naive Bayes
Gaussian Naive Bayes is a classification algorithm based on Bayes' theorem with the assumption that the features follow a Gaussian distribution. It is particularly effective for continuous data.

In [11]:
from sklearn.naive_bayes import GaussianNB

gnb_model = GaussianNB()
gnb_model.fit(X_train, Y_train)
Y_pred_gnb = gnb_model.predict(X_test)
gnb_accuracy = accuracy_score(Y_test, Y_pred_gnb)
print("Training Accuracy Score:", gnb_accuracy)

Training Accuracy Score: 0.9920787304375053


# Conclusion
In conclusion we find out that Logistic Regression (with L1 Regularization) Algorithm and Gaussian Naive Bayes Algorithm gives an accuracy of 99.55% and 99.20% respectively.